terra_notebook_utils/table.py

```diff
@@ -1,10 +1,131 @@
 from firecloud import fiss
-from typing import Optional
 
 from terra_notebook_utils import WORKSPACE_GOOGLE_PROJECT, WORKSPACE_NAME
+from terra_notebook_utils.utils import _AsyncContextManager
+
+
+Row = namedtuple("Row", "name attributes")
+
+VALUE = Union[str, int, float, bool]
+ATTRIBUTES = Dict[str, Union[VALUE, Iterable[VALUE]]]
+UPDATE_OPS = List[Dict[str, Any]]
+COLUMN_HEADERS = Tuple[str, ...]
+ROW_LIKE = Tuple[str, ATTRIBUTES]
+
+class Writer(_AsyncContextManager):
+    """
+    Distribute row uploads across as few API calls as possible.
+    Uploads are performed in the background.
+    Also transparantly handles sequences, which the Firecloud API makes difficult. (These must be uploaded and modified
+    via separate API calls.)
+    """
+    def __init__(self, name: str, **kwargs):
+        self.name = name
+        self._init_request_data()
+        self._workspace_name = kwargs.get("workspace_name", WORKSPACE_NAME)
+        self._workspace_google_project = kwargs.get("workspace_google_project", WORKSPACE_GOOGLE_PROJECT)
+
+    def _init_request_data(self):
+        self._tsvs: Dict[COLUMN_HEADERS, str] = defaultdict(list)
+        self._row_update_request_data: Dict[COLUMN_HEADERS, List[Tuple[Row, UPDATE_OPS]]] = dict()
+
+    def _get_row_update_request_data(self, row: Row) -> List[Dict[str, Any]]:
+        request_data = list()
+        for name, val in row.attributes.items():
+            if isinstance(val, str):
+                update_ops = list()
+            elif isinstance(val, (int, float, bool)):
+                update_ops = [dict(op="AddUpdateAttribute", attributeName=name, addUpdateAttribute=val)]
+            elif hasattr(val, "__iter__"):
+                update_ops = [dict(op="RemoveAttribute", attributeName=name)]
+                types: Set[type] = set()
+                for m in val:
+                    update_ops.append(dict(op="AddListMember", attributeListName=name, newMember=m))
+                    types.add(type(m))
+                assert 1 == len(types)
+# assert types.pop() in VALUE.__args__, f"Row: {row.name}: List members must be of type {VALUE.__args__}"
+            else:
+                pass
+                # raise TypeError(f"Row {row.name}: data table value must be of type{VALUE.__args__}")
+            request_data.extend(update_ops)
+        return request_data
+
+    def put_row(self, item: Union[ROW_LIKE, ATTRIBUTES]) -> str:
+        if isinstance(item, dict):
+            row = Row(f"{uuid4()}", item)
+        else:
+            row = Row(*item)
+        column_headers = tuple(sorted(row.attributes.keys()))
+        if column_headers not in self._tsvs:
+            self._tsvs[column_headers] = "\t".join([f"{self.name}_id", *column_headers])
+            self._row_update_request_data[column_headers] = list()
+        self._tsvs[column_headers] += (
+            os.linesep
+            + row.name + "\t"
+            + "\t".join(row.attributes[c] if isinstance(row.attributes[c], str)
```

Comment:
What is the `x` default?  Is it possible this could be an `int` or is this only handling the `None` case?

Reply:
Only string values can be uploaded in the original TSV upload. Other types require additional Firecloud API calls. “x” is a placeholder value.

Will add comment.